core/executor/src/wasm_executor.rs

```diff
@@ -124,1201 +270,1066 @@ fn deadline_to_timestamp(deadline: u64) -> Option<offchain::Timestamp> {
+				context.write_memory(value_data, &value[..written])
+					.map_err(|_| "Invalid attempt to set value in ext_get_child_storage_into")?;
+				Ok(value.len() as u32)
 			} else {
-				"<empty>"
-			},
-			HexDisplay::from(&key),
-		);
+				Ok(u32::max_value())
+			}
+		}
+
+		ext_storage_root(result: Pointer<u8>) {
+			let r = runtime_io::storage_root();
+			context.write_memory(result, r.as_ref())
+				.map_err(|_| "Invalid attempt to set memory in ext_storage_root")?;
+			Ok(())
+		}
 
-		if let Some(value) = maybe_value {
-			let offset = this.heap.allocate(value.len() as u32)? as u32;
-			this.memory.set(offset, &value)
-				.map_err(|_| "Invalid attempt to set memory in ext_get_allocated_storage")?;
-			this.memory.write_primitive(written_out, value.len() as u32)
-				.map_err(|_| "Invalid attempt to write written_out in ext_get_allocated_storage")?;
+		ext_child_storage_root(
+			storage_key_data: Pointer<u8>,
+			storage_key_len: WordSize,
+			written_out: Pointer<u32>,
+		) -> Pointer<u8> {
+			let storage_key = context.read_memory(storage_key_data, storage_key_len)
+				.map_err(|_| "Invalid attempt to determine storage_key in ext_child_storage_root")?;
+			let value = runtime_io::child_storage_root(&storage_key);
+
+			let offset = context.allocate_memory(value.len() as u32)?;
+			context.write_memory(offset, &value)
+				.map_err(|_| "Invalid attempt to set memory in ext_child_storage_root")?;
+			context.write_primitive(written_out, value.len() as u32)
+				.map_err(|_| "Invalid attempt to write written_out in ext_child_storage_root")?;
 			Ok(offset)
-		} else {
-			this.memory.write_primitive(written_out, u32::max_value())
-				.map_err(|_| "Invalid attempt to write failed written_out in ext_get_allocated_storage")?;
-			Ok(0)
 		}
-	},
-	// return 0 and place u32::max_value() into written_out if no value exists for the key.
-	ext_get_allocated_child_storage(
-		storage_key_data: *const u8,
-		storage_key_len: u32,
-		key_data: *const u8,
-		key_len: u32,
-		written_out: *mut u32
-	) -> *mut u8 => {
-		let storage_key = this.memory.get(
-			storage_key_data,
-			storage_key_len as usize
-		).map_err(|_| "Invalid attempt to determine storage_key in ext_get_allocated_child_storage")?;
-		let key = this.memory.get(
-			key_data,
-			key_len as usize
-		).map_err(|_| "Invalid attempt to determine key in ext_get_allocated_child_storage")?;
-
-		let maybe_value = {
-			let storage_key = ChildStorageKey::from_slice(&storage_key)
-				.ok_or_else(|| "ext_get_allocated_child_storage: child storage key is not valid")?;
-			this.ext.child_storage(storage_key, &key)
-		};
 
-		debug_trace!(
-			target: "wasm-trace", "*** Getting child storage: {} -> {} == {}   [k={}]",
-			primitives::hexdisplay::ascii_format(&storage_key),
-			if let Some(_preimage) = this.hash_lookup.get(&key) {
-				format!("%{}", ::primitives::hexdisplay::ascii_format(&_preimage))
+		ext_storage_changes_root(
+			parent_hash_data: Pointer<u8>,
+			_len: WordSize,
```

Comment:
Why don't we check for the length anymore?

Reply:
We could readd the check, but for all the other places where we take fixed size data types we don't have this check. If you look into the `sr-io` api that calls this function, it only takes an array of 32 elements.

Reply:
Fine by me! Although in this case we should remove `len` param.

Reply:
Yeah, but this would break Kusama as we could not import the wasm blob anymore.

Reply:
Let's leave a TODO then?